pkg/nfd-worker/nfd-worker.go

```diff
@@ -190,11 +190,19 @@ func (i *infiniteTicker) Reset(d time.Duration) {
 
 // Run feature discovery.
 func (w *nfdWorker) runFeatureDiscovery() error {
+	discoveryStart := time.Now()
 	for _, s := range w.featureSources {
-		klog.V(2).Infof("running discovery for %q source", s.Name())
+		currentSourceStart := time.Now()
 		if err := s.Discover(); err != nil {
 			klog.Errorf("feature discovery of %q source failed: %v", s.Name(), err)
 		}
+		klog.V(2).Infof("discovery duration for %q source: %v", s.Name(), time.Since(currentSourceStart))
+	}
+
+	discoveryDuration := time.Since(discoveryStart)
+	klog.V(2).Infof("feature discovery for sources lasted for: %v", discoveryDuration)
+	if discoveryDuration > 5*time.Second {
+		klog.Warningf("feature discovery for sources lasted for more than 5s (%v)", discoveryDuration)
```

Comment:
Warning at 5s is not useful when the discovery interval is smaller.  I would go for something like this (untested!):
```suggestion
	if discoveryDuration > core.sleepInterval/2 {
		klog.Warningf("Feature discovery sources took over half (%v) of sleep interval (%v)", discoveryDuration, core.sleepInterval)
```

(Other alternative would be having measurement and warning like this where the sleep is done instead.)

PS. commit message is missing "Fixes:" tag for https://github.com/kubernetes-sigs/node-feature-discovery/issues/899.

Reply:
@VillePihlava please address, I agree with @eero-t

Reply:
@eero-t PTAL

Reply:
Also added a check to see if the duration is positive (negative durations should not cause a warning)

Reply:
Good point, looks fine now!